lib/fog/vsphere/models/compute/server.rb

```diff
@@ -348,15 +354,28 @@ def initialize_interfaces
           end
         end
 
+        def unassigned_volumes?
+          attributes[:volumes]&.any? { |vol| !vol.key?(:controller_key) } || false
+        end
+
+        def update_controller_key(vol)
+          case vol
+          when Hash
+            vol[:controller_key] ||= attributes[:scsi_controllers].first&.key || 1000
+          when Object
+            vol.controller_key ||= attributes[:scsi_controllers].first&.key || 1000
+          end
+        end
+
         def initialize_volumes
-          if attributes[:volumes] && attributes[:volumes].is_a?(Array)
-            attributes[:volumes].map! do |vol|
-              if vol.is_a?(Hash)
-                service.volumes.new({ server: self }.merge(vol))
-              else
-                vol.server = self
-                vol
-              end
+          return unless attributes[:volumes].is_a?(Array)
+          attributes[:volumes].map! do |vol|
+            update_controller_key(vol)
+            if vol.is_a?(Hash)
+              service.volumes.new({ server: self }.merge(vol))
+            else
+              vol.server = self
+              vol
```

Comment:
If you reorder the calls to:
```suggestion
            if vol.is_a?(Hash)
              vol = service.volumes.new({ server: self }.merge(vol))
            end
            vol.server = self
            update_controller_key(vol)         
            vol
```

You will be able to remove the `case` statement in `update_controller_key`:
``` ruby
        def update_controller_key(vol)
            vol.controller_key ||= attributes[:scsi_controllers].first&.key || 1000
        end
```

because `vol` will always contain an object.

Reply:
Interesting. Thanks Shim, updated :)